### ISSUE TITLE
[testing] configurable flush

### ISSUE THREAD

**Issue by bashmish** [open-wc/open-wc#32]
Currently `flush` is hardcoded in the fixture helpers implementation and based on the old WCT API.

https://github.com/open-wc/open-wc/blob/2b06f4770e4881141172c37fb4f7e6323da7bccc/packages/testing/helpers.js#L92-L104

I don't think it is helpful to pull the legacy of WCT here.

How about making it fully configurable and provide a clear API for that?

If we still want to have out-of-the-box behavior, we can keep `setTimeout` to delay till next task by default, or just use a simple `Promise` to delay till next microtask by default.

**Comment by LarsDenBakker** [comment]
Why is a flush needed at all? I've been using:

```
export function waitForRender() {
  return new Promise((resolve) => {
    requestAnimationFrame(() => {
      setTimeout(() => {
        resolve();
      });
    });
  });
}
```
It's semantically correct: RAF means before the next render, and a timeout ensures the code is executed after the next render. I use it in any code that relies on any kind of rendering to complete, both production code and tests, and it hasn't failed me anywhere.

**Comment by bashmish** [comment]
@LarsDenBakker everywhere in our code even just simple `setTimeout` always worked. In Polymer's tests they use only `requestAnimationFrame` and also works for them everythere (https://github.com/Polymer/lit-element/blob/master/src/test/test-helpers.ts#L22).

This is sort of a hacky abstraction and depends on the implementation of the base class.

**Comment by bashmish** [comment]
@LarsDenBakker makes sense

Just a side note, it does not mean we should go that deep. As far as I understand the trend in React and async rendering there (aka React Fibers, which is not anymore marketed as that) they want to be able to pause rendering which is not able to be finished within one frame and postpone part of the job till next frame. So the final rendered DOM might be ready only after two frames, or in general in any number of frames, so it is sort of non-deterministic. Does not necessarily apply to testing environment as it will probably be a bad idea to have this non-deterministic behavior in tests anyway :D not to say that this is so far only React which is going in that direction and is still not there yet.

So, should be fine to change the flush here to the same implementation as here https://github.com/Polymer/lit-element/blob/master/src/test/test-helpers.ts#L22 , and maybe rename to `nextFrame` as well (`await nextFrame()` will look a way better to me).

**Comment by LarsDenBakker** [comment]
Youre right, in that case it will not be sufficient indeed.

**Comment by daKmoR** [comment]
>I don't think it is helpful to pull the legacy of WCT here.
> How about making it fully configurable and provide a clear API for that?

it sort of is that right? so if `window.flush` is already defined - e.g. while using WCT it will use that.

or you can easily define your own like so:
```js
window.flush = () => new Promise((resolve) => requestAnimationFrame(resolve));
```

by default, it's on the safe side and waits for 2 full tasks to be absolutely sure everything is done (or is the question about changeing the default?)

**Comment by thepassle** [comment]
What's the consensus on this? Remove the `flush` wct legacy, and add `waitForRender`?

**Comment by bashmish** [comment]
@thepassle I've used `nextFrame` recently pretty much everywhere in tests written during this year (and also with all possible configurations including WCT, karma, wallaby.js) and it worked just well. I thought the consensus was as explained in my previous comment.

**Comment by daKmoR** [comment]
agreed let's remove it 👍 

will also need to update timings documentation to properly explain aTimeout and nextFrame

**Comment by daKmoR** [comment]
remove and therefore fixed in the latest release 🎉
